protoc-gen-kotlin/protoc-gen-kotlin-common/src/main/kotlin/pbandk/gen/CodeGenerator.kt

```diff
@@ -27,14 +27,18 @@ open class CodeGenerator(val file: File, val kotlinTypeMappings: Map<String, Str
     }
 
     protected fun writeEnumType(type: File.Type.Enum) {
-        // Enums are data classes w/ a single value and a companion object with known values
-        line().line("data class ${type.kotlinTypeName}(override val value: Int) : pbandk.Message.Enum {").indented {
-            line("companion object : pbandk.Message.Enum.Companion<${type.kotlinTypeName}> {").indented {
-                type.values.forEach { line("val ${it.kotlinValueName} = ${type.kotlinTypeName}(${it.number})") }
+        // Enums are data classes w/ a value and a name, and a companion object with known values
+        line().line("data class ${type.kotlinTypeName}(override val value: Int, override val name: String) : pbandk.Message.NamedEnum {").indented {
```

Comment:
❌ One thing to be careful about is that protobuf enums allow aliases (i.e. multiple names with the same numeric value). The protobuf spec says that all aliases should be considered equal to each other. A Kotlin data class will have an `equals()` implementation that compares all constructor properties of the data class for equality. So with this additional constructor property, aliases will no longer be equal.

There are some other default behaviors of Kotlin data classes that also don't make sense for a protobuf enum (e.g. the generated `copy()` method), so I think overall it makes more sense to not use a `data class` here and just generate whichever methods we need.

I can think of a couple solutions:

1. Make the enum a plain `class` instead of a `data class` and generate our own `equals()` method.

2. The more I think about it, I'm wondering if maybe a `sealed class` would be a better fit here rather than just generating a constant for each enum value. Something like:

```kt
/* proto definition:
enum SomeEnum {
    INVALID = 0;
    FOO = 1;
    BAR = 2;
}
*/

sealed class SomeEnum(override val value: Int, override val name: String? = null) : pbankd.Message.Enum {
    override fun equals(other: Any?) = other is SomeEnum && other.value == value
    override fun hashCode() = value.hashCode()
    override fun toString() = "SomeEnum.${name ?: "UNRECOGNIZED"}(value=$value)"

    object Invalid : SomeEnum(0, "INVALID")
    object Foo : SomeEnum(1, "FOO")
    object Bar : SomeEnum(2, "BAR")
    object Unrecognized(value: Int) : SomeEnum(value)

    companion object : pbandk.Message.Enum.Companion<SomeEnum> {
        override val values = listOf(Invalid, Foo, Bar)
        override fun fromValue(value: Int) = values.firstOrNull { it.value == value } ?: Unrecognized(value)
        override fun fromName(name: String) = values.firstOrNull { it.name == name } ?: throw IllegalArgumentException("No SomeEnum with name: $name")
    }
}
```

Reply:
Cool, I'll do this, and strip the prefix, before merging.

Reply:
Also, [here](https://github.com/streem/pb-and-k/blob/master/runtime/runtime-common/src/main/kotlin/pbandk/wkt/descriptor.kt#L118) is one example of an existing `Enum`, which I plan to leave alone, and not rewrite to a `sealed class`... i'll just do this strategy for `NamedEnum`s

Reply:
Oh I see what you're referring to now. Those files are actually all auto-generated from the relevant `.proto` files. There's a bit of a bootstrapping problem however since `pb-and-k` uses those files during the code generation process, but those files are generated by `pb-and-k` applied to the built-in `.proto` files that ship with protobuf.

I suppose we can have the separate `NamedEnum` interface as a temporary stop-gap to allow this PR to merge. We do need to re-generate those files at some point. Once we do, I think `NamedEnum` should be merged back with `Enum`. Those files don't yet include some other changes I've made to the code generator either. I started looking at how to do this the other day but haven't quite figured out the right set of bootstrapping commands to run.

BTW, please also update the `README.md` enum example with the new generated code.

Reply:
README updated

Reply:
After plugging in everywhere, not a huge fan of things like `Room.Status.Closed`, as opposed to `Room.Status.CLOSED`... wish we could have the proper capitalized enums

Reply:
`Room.Status.Closed` is similar to how an enum value looks in Swift. And since sealed classes are a popular aspect of Kotlin, it doesn't look out of place to me in a Kotlin codebase. Does it look strange to you in Kotlin because you're used to seeing Java-style enums?

I'm personally a fan of `Room.Status.Closed`. But I'm not opposed to slightly breaking Kotlin capitalization rules here and outputting the `object` names as `CLOSED` instead of `Closed`. Since we are emulating an enum at the end of the day, I don't think the all-caps style will look that out of place in code that's using this.